src/app/OperationalDeviceProxy.cpp

```diff
@@ -144,21 +145,31 @@ CHIP_ERROR OperationalDeviceProxy::EstablishConnection()
 {
     // Create a UnauthenticatedSession for CASE pairing.
     // Don't use mSecureSession here, because mSecureSession is for encrypted communication.
+    VerifyOrReturnError(mCASESession == nullptr, CHIP_ERROR_INCORRECT_STATE);
     Optional<SessionHandle> session = mInitParams.sessionManager->CreateUnauthenticatedSession(mDeviceAddress);
     VerifyOrReturnError(session.HasValue(), CHIP_ERROR_NO_MEMORY);
 
     session.Value().GetUnauthenticatedSession()->SetMRPIntervals(mMrpIdleInterval, mMrpActiveInterval);
 
-    Messaging::ExchangeContext * exchange = mInitParams.exchangeMgr->NewContext(session.Value(), &mCASESession);
+    mCASESession = chip::Platform::New<CASESession>();
+    VerifyOrReturnError(mCASESession != nullptr, CHIP_ERROR_NO_MEMORY);
+    auto deferred = MakeDefer([&]() {
+        if (mState != State::Connecting)
+        {
+            FreeCASESession();
+        }
+    });
```

Comment:
The real solution is to make CASESession smaller, it should be less than 128 bytes based on the necessary session context. What accounts for the 1.4K?

The issue with putting it in heap are:

- Devices with lower heap space may run-time fail to allocate, which would be really hard to account
- On small memory devices, 1.4K is actually a possibly hard-to-allocate block even if there is > 1.4K left, due to fragmentation (this issue arose on many of the platforms with < 128K RAM before).
- There are already requirements in the spec for a minimum number of CASE sessions contexts to be supported, and clear rules for eviction in case of running out of memory. We should try to follow that rather than expect the CASE session to fit, and allocating it to heap. In fact, it's likely better to have a pool of them, and a management strategy for the eviction.

Reply:
I agree, the size of the CASESession should be optimized.

The CASESession object is mainly used during the session establishment phase. Once the session is established, it is used for serializing and caching to enable session resumption. From that point of view, the object can be added to cache (as the serialized object) and freed.

Using heap does complicate the requirement for a device to support at least some number of CASE sessions. So, optimizing the size of the object would be more ideal from that point of view.